virtual_ecosystem/core/core_components.py

```diff
@@ -218,33 +272,120 @@ def __post_init__(self, config: Config) -> None:
 
         # Validate configuration
         self.canopy_layers = _validate_positive_integer(lyr_config["canopy_layers"])
-        self.soil_layers = _validate_soil_layers(lyr_config["soil_layers"])
+        self.soil_layers = np.array(_validate_soil_layers(lyr_config["soil_layers"]))
 
         # Other heights should all be positive floats
         for attr, value in (
             ("above_canopy_height_offset", lyr_config["above_canopy_height_offset"]),
             ("surface_layer_height", lyr_config["surface_layer_height"]),
-            ("subcanopy_layer_height", lyr_config["subcanopy_layer_height"]),
         ):
             setattr(self, attr, _validate_positive_finite_numeric(value, attr))
 
-        self.layer_roles = (
+        # Get the layer role sequence
+        self.layer_roles = np.array(
             ["above"]
             + ["canopy"] * int(self.canopy_layers)
-            + ["subcanopy"]
             + ["surface"]
-            + ["soil"] * len(self.soil_layers)
+            + ["topsoil"]
+            + ["subsoil"] * (len(self.soil_layers) - 1)
         )
 
+        # Record the number of layers and layer indices
         self.n_layers = len(self.layer_roles)
+        self.layer_indices = np.arange(0, self.n_layers)
+
+        # Set up the various indices onto those roles
+        self.role_indices_bool = {
+            layer_role: self.layer_roles == layer_role
+            for layer_role in ("above", "canopy", "surface", "topsoil", "subsoil")
+        }
+        self.role_indices = {
+            ky: np.nonzero(vl)[0] for ky, vl in self.role_indices_bool.items()
+        }
+
+        # Add the `all_soil` and `atmospheric` indices
+        self.role_indices["all_soil"] = np.concatenate(
+            [self.role_indices["topsoil"], self.role_indices["subsoil"]]
+        )
```

Comment:
Maybe, for consistency with the atmosphere - and because I think it looks like a nicer syntax - I would move this below and do:
```suggestion
        self.role_indices["all_soil"] = np.where(
            self.role_indices_bool["all_soil"]
        )[0]
```

Reply:
Updated in #451